src/lib/audio/audio-util.js

```diff
@@ -59,9 +60,39 @@ const encodeAndAddSoundToVM = function (vm, samples, sampleRate, name, callback)
     });
 };
 
+const downsampleIfNeeded = (samples, sampleRate, resampler) => {
+    const duration = samples.length / sampleRate;
+    const encodedByteLength = samples.length * 2; /* bitDepth 16 bit */
+    // Resolve immediately if already within byte limit
+    if (encodedByteLength < SOUND_BYTE_LIMIT) {
+        return Promise.resolve({samples, sampleRate});
+    }
+    // If encodeable at 22khz, resample and call submitNewSamples again
+    if (duration * 22050 * 2 < SOUND_BYTE_LIMIT) {
+        return resampler({samples, sampleRate}, 22050);
+    }
+    // Cannot save this sound at 22khz, refuse to edit
+    // In the future we could introduce further compression here
+    return Promise.reject('Sound too large to save, refusing to edit');
+};
+
+const backupDownSampler = (buffer, newRate) => {
```

Comment:
I'm not sure I understand what "backup" means in this context. I know from your perspective this downsampling is abnormal, but to an outsider like me it doesn't seem that way -- this just seems like more downsampling to make the data smaller and more editable.

Reply:
Ah yeah. So it is the "backup" method because the ideal method is to use the audio context to resample the audio, which involves some interpolation instead of just dropping every other sample, which introduces audio artifacts. But if the audio context does not support resampling at the desired frequency (like safari with anything other than 44khz), this is the "backup" option. @ericrosenbaum I think adding this as jsdoc (and generally jsdoc-ing the arguments to all these utility functions) would be a good idea

Reply:
I see -- there are two types of downsampling -- you might think of them as:
1)"downsampleUsingInterpolation" or maybe "downsampleSmartly" or "smartDownsample", and
2) "downsampleByDroppingSamples" or "dropEveryOtherSample" or "bruteForceDownsample".

I think drawing this distinction using a pair of functions with linked names might make their respective roles clearer. For me, the "backup" nature of 2) seems like it would be clearer if it came in the context of the branching logic, with the function name being more specific to what that function does to its data (rather than the name referring to the situation in which it is called). 

But, my function naming preferences aren't necessarily better -- I'm fine with what you prefer, as long as you consider a change like i'm suggesting.